drivers/sensor/ags10/ags10.c

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright (c) 2023 Balthazar Deliers
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT aosong_ags10
+
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/byteorder.h>
+
+#include "ags10.h"
+
+LOG_MODULE_REGISTER(AGS10, CONFIG_SENSOR_LOG_LEVEL);
+
+static int ags10_write(const struct device *dev, uint8_t cmd, uint8_t* data, uint8_t bytes)
+{
+	const struct ags10_config *conf = dev->config;
+	uint8_t send_buf[6] = {0};
+
+	send_buf[0] = cmd;
+	memcpy(&send_buf[1], data, bytes);
+
+	return i2c_write_dt(&conf->bus, send_buf, bytes);
+}
+
+static int ags10_read(const struct device *dev, uint8_t cmd, uint8_t *data, uint8_t rx_bytes)
+{
+	const struct ags10_config *conf = dev->config;
+
+	uint8_t recv_buf[5] = {0};
+	int ret = i2c_write_read_dt(&conf->bus,
+				    &cmd,
+				    sizeof(cmd),
+				    &recv_buf,
+				    rx_bytes);
+	if (ret < 0) {
+		return ret;
+	}
+
+	memcpy(data, recv_buf, rx_bytes);
+
+	return 0;
+}
+
+/*
+ * @brief Calculation provided by the manufacturer in the datasheet part 3.7
+ * @param data  Data buffer on which the CRC needs to be calculated
+ * @param size  Data size in bytes
+ */
+static uint8_t ags10_crc(uint8_t *data, uint8_t size)
+{
+    uint8_t i,byte,crc = 0xFF;
+
+    for(byte= 0; byte < size; byte++)
+    {
+        crc ^= (data[byte]);
+        for(i=0;i<8;i++)
+        {
+            if (crc & 0x80) { crc = (crc << 1) ^ 0x31; }
+            else { crc = (crc << 1); }
+        }
+    }
+    return crc;
+}
+
+static int ags10_sample_fetch(const struct device *dev,
+				 enum sensor_channel chan)
+{
+	struct ags10_data *data = dev->data;
+	int ret = -ENOTSUP;
+
+	if (chan == SENSOR_CHAN_VOC || chan == SENSOR_CHAN_ALL) {
+
+        uint8_t recv_buf[5] = {0};
+
+        ret = ags10_read(dev, AGS10_CMD_DATA_ACQUISITION, recv_buf, 5);
+
+        if (ret == 0)
+        {
+            /* If CRC is valid and data is valid too */
+            if (ags10_crc(&recv_buf[0], 4) == recv_buf[4]
+                && ((recv_buf[0] & AGS10_MSK_STATUS) == AGS10_REG_STATUS_NRDY_READY))
+            {
+                data->status   = recv_buf[0] & AGS10_MSK_STATUS;
+                data->tvoc_ppb = sys_get_be24(&recv_buf[1]);
+                return 0;
+            }
+            else
+            {
+                LOG_WRN("Bad CRC or data not ready");
+                return -EIO;
+            }
+        }
+	}
+
+    return ret;
+}
+
+static int ags10_channel_get(const struct device *dev,
+				enum sensor_channel chan,
+				struct sensor_value *val)
+{
+	struct ags10_data *data = dev->data;
+
+	if (chan == SENSOR_CHAN_VOC) {
+		val->val1 = (data->tvoc_ppb > 0xFFFF) ? 0xFFFF : data->tvoc_ppb; // Clamp data

```

Comment:
It looks like this is a 24-bit value based on line 88, does the sensor not exceed 65535?

Reply:
The sensor can go up to 99999 but in the zephyr API the voc channel is described to be on an uint16, ths is why I clamp the value

Reply:
Indeed, I first clamped it to an uint16 as on ccs811 for instance, but I changed it back to the full range